r/aws_vpc_peering_connection_options: Only modify VPC Peering options that have actually changed

ewbankkit opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#pull-requests --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes https://github.com/terraform-providers/terraform-provider-aws/issues/12113.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
resource/aws_vpc_peering_connection_options: Only modify options that have changed
```

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcPeeringConnectionOptions_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 2 -run=TestAccAWSVpcPeeringConnectionOptions_ -timeout 120m
=== RUN   TestAccAWSVpcPeeringConnectionOptions_basic
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_basic
=== RUN   TestAccAWSVpcPeeringConnectionOptions_sameRegionSameAccountUpdates
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_sameRegionSameAccountUpdates
=== RUN   TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountBasic
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountBasic
=== RUN   TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountUpdates
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountUpdates
=== RUN   TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountBasic
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountBasic
=== RUN   TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountUpdates
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountUpdates
=== CONT  TestAccAWSVpcPeeringConnectionOptions_basic
=== CONT  TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountUpdates
--- PASS: TestAccAWSVpcPeeringConnectionOptions_basic (41.32s)
=== CONT  TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountBasic
--- PASS: TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountUpdates (73.05s)
=== CONT  TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountUpdates
--- PASS: TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountBasic (45.75s)
=== CONT  TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountBasic
--- PASS: TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountUpdates (69.23s)
=== CONT  TestAccAWSVpcPeeringConnectionOptions_sameRegionSameAccountUpdates
--- PASS: TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountBasic (46.71s)
--- PASS: TestAccAWSVpcPeeringConnectionOptions_sameRegionSameAccountUpdates (64.91s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	169.686s
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVPCPeeringConnection_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 2 -run=TestAccAWSVPCPeeringConnection_ -timeout 120m
=== RUN   TestAccAWSVPCPeeringConnection_basic
=== PAUSE TestAccAWSVPCPeeringConnection_basic
=== RUN   TestAccAWSVPCPeeringConnection_plan
=== PAUSE TestAccAWSVPCPeeringConnection_plan
=== RUN   TestAccAWSVPCPeeringConnection_tags
=== PAUSE TestAccAWSVPCPeeringConnection_tags
=== RUN   TestAccAWSVPCPeeringConnection_options
=== PAUSE TestAccAWSVPCPeeringConnection_options
=== RUN   TestAccAWSVPCPeeringConnection_failedState
=== PAUSE TestAccAWSVPCPeeringConnection_failedState
=== RUN   TestAccAWSVPCPeeringConnection_peerRegionAutoAccept
=== PAUSE TestAccAWSVPCPeeringConnection_peerRegionAutoAccept
=== RUN   TestAccAWSVPCPeeringConnection_region
=== PAUSE TestAccAWSVPCPeeringConnection_region
=== RUN   TestAccAWSVPCPeeringConnection_accept
=== PAUSE TestAccAWSVPCPeeringConnection_accept
=== RUN   TestAccAWSVPCPeeringConnection_optionsNoAutoAccept
=== PAUSE TestAccAWSVPCPeeringConnection_optionsNoAutoAccept
=== CONT  TestAccAWSVPCPeeringConnection_basic
=== CONT  TestAccAWSVPCPeeringConnection_peerRegionAutoAccept
--- PASS: TestAccAWSVPCPeeringConnection_peerRegionAutoAccept (21.17s)
=== CONT  TestAccAWSVPCPeeringConnection_optionsNoAutoAccept
--- PASS: TestAccAWSVPCPeeringConnection_basic (39.62s)
=== CONT  TestAccAWSVPCPeeringConnection_accept
--- PASS: TestAccAWSVPCPeeringConnection_optionsNoAutoAccept (24.94s)
=== CONT  TestAccAWSVPCPeeringConnection_region
--- PASS: TestAccAWSVPCPeeringConnection_region (38.97s)
=== CONT  TestAccAWSVPCPeeringConnection_options
--- PASS: TestAccAWSVPCPeeringConnection_accept (84.16s)
=== CONT  TestAccAWSVPCPeeringConnection_failedState
--- PASS: TestAccAWSVPCPeeringConnection_failedState (21.11s)
=== CONT  TestAccAWSVPCPeeringConnection_tags
--- PASS: TestAccAWSVPCPeeringConnection_options (64.61s)
=== CONT  TestAccAWSVPCPeeringConnection_plan
--- PASS: TestAccAWSVPCPeeringConnection_tags (38.84s)
--- PASS: TestAccAWSVPCPeeringConnection_plan (34.74s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	184.509s
```

Without the fix, updating VPC peering connection options for cross-region peerings would fail:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcPeeringConnectionOptions_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 1 -run=TestAccAWSVpcPeeringConnectionOptions_ -timeout 120m
=== RUN   TestAccAWSVpcPeeringConnectionOptions_basic
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_basic
=== RUN   TestAccAWSVpcPeeringConnectionOptions_sameRegionSameAccountUpdates
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_sameRegionSameAccountUpdates
=== RUN   TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountBasic
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountBasic
=== RUN   TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountUpdates
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountUpdates
=== RUN   TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountBasic
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountBasic
=== RUN   TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountUpdates
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountUpdates
=== CONT  TestAccAWSVpcPeeringConnectionOptions_basic
--- PASS: TestAccAWSVpcPeeringConnectionOptions_basic (41.92s)
=== CONT  TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountBasic
--- PASS: TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountBasic (46.27s)
=== CONT  TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountUpdates
--- PASS: TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccountUpdates (70.79s)
=== CONT  TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountBasic
--- PASS: TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountBasic (46.41s)
=== CONT  TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountUpdates
--- FAIL: TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccountUpdates (56.62s)
    testing.go:654: Step 2 error: errors during apply:
        
        Error: error modifying VPC Peering Connection (pcx-0db05882d9de2b5fb) Options: InvalidParameterValue: Requester’s VPC Peering connection options cannot be modified for a different region
        	status code: 400, request id: db4628d4-5c79-496f-911b-303cf2e58108
        
          on /tmp/tf-test059046709/main.tf line 63:
          (source code not available)
        
        
        
        Error: error modifying VPC Peering Connection (pcx-0db05882d9de2b5fb) Options: InvalidParameterValue: Accepter’s VPC Peering connection options cannot be modified for a different region
        	status code: 400, request id: ef53e7fb-12a3-480d-a02b-e9fb86e190c6
        
          on /tmp/tf-test059046709/main.tf line 52:
          (source code not available)
        
        
=== CONT  TestAccAWSVpcPeeringConnectionOptions_sameRegionSameAccountUpdates
--- PASS: TestAccAWSVpcPeeringConnectionOptions_sameRegionSameAccountUpdates (65.81s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	327.905s
FAIL
GNUmakefile:25: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

relnetops commented:
@ewbankkit Thanks for this PR.  This is currently affecting one of our peering connections.  It's been a few months.  Can we get this reviewed?

ewbankkit commented:
Rebased to fix merge conflict.
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVPCPeeringConnection_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 2 -run=TestAccAWSVPCPeeringConnection_ -timeout 120m
=== RUN   TestAccAWSVPCPeeringConnection_basic
=== PAUSE TestAccAWSVPCPeeringConnection_basic
=== RUN   TestAccAWSVPCPeeringConnection_plan
=== PAUSE TestAccAWSVPCPeeringConnection_plan
=== RUN   TestAccAWSVPCPeeringConnection_tags
=== PAUSE TestAccAWSVPCPeeringConnection_tags
=== RUN   TestAccAWSVPCPeeringConnection_options
=== PAUSE TestAccAWSVPCPeeringConnection_options
=== RUN   TestAccAWSVPCPeeringConnection_failedState
=== PAUSE TestAccAWSVPCPeeringConnection_failedState
=== RUN   TestAccAWSVPCPeeringConnection_peerRegionAutoAccept
=== PAUSE TestAccAWSVPCPeeringConnection_peerRegionAutoAccept
=== RUN   TestAccAWSVPCPeeringConnection_region
=== PAUSE TestAccAWSVPCPeeringConnection_region
=== RUN   TestAccAWSVPCPeeringConnection_accept
=== PAUSE TestAccAWSVPCPeeringConnection_accept
=== RUN   TestAccAWSVPCPeeringConnection_optionsNoAutoAccept
=== PAUSE TestAccAWSVPCPeeringConnection_optionsNoAutoAccept
=== CONT  TestAccAWSVPCPeeringConnection_basic
=== CONT  TestAccAWSVPCPeeringConnection_peerRegionAutoAccept
--- PASS: TestAccAWSVPCPeeringConnection_peerRegionAutoAccept (21.43s)
=== CONT  TestAccAWSVPCPeeringConnection_optionsNoAutoAccept
--- PASS: TestAccAWSVPCPeeringConnection_basic (40.14s)
=== CONT  TestAccAWSVPCPeeringConnection_accept
--- PASS: TestAccAWSVPCPeeringConnection_optionsNoAutoAccept (24.86s)
=== CONT  TestAccAWSVPCPeeringConnection_region
--- PASS: TestAccAWSVPCPeeringConnection_region (39.26s)
=== CONT  TestAccAWSVPCPeeringConnection_options
--- PASS: TestAccAWSVPCPeeringConnection_accept (85.58s)
=== CONT  TestAccAWSVPCPeeringConnection_failedState
--- PASS: TestAccAWSVPCPeeringConnection_failedState (21.42s)
=== CONT  TestAccAWSVPCPeeringConnection_tags
--- PASS: TestAccAWSVPCPeeringConnection_options (64.76s)
=== CONT  TestAccAWSVPCPeeringConnection_plan
--- PASS: TestAccAWSVPCPeeringConnection_plan (34.39s)
--- PASS: TestAccAWSVPCPeeringConnection_tags (39.30s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	186.521s
```

ewbankkit commented on aws/resource_aws_vpc_peering_connection_test.go:
```diff
@@ -509,6 +509,7 @@ func testAccCheckAWSVpcPeeringConnectionOptionsWithProvider(n, block string, opt
 }
 
 func TestAccAWSVPCPeeringConnection_peerRegionAutoAccept(t *testing.T) {
+	var providers []*schema.Provider
```
Else get errors like

```
    === CONT  TestAccAWSVPCPeeringConnection_peerRegionAutoAccept
        testing.go:677: Step 0, expected error:
    
            Error initializing context: 2 problems:
    
            - Could not satisfy plugin requirements:
            Plugin reinitialization required. Please run "terraform init".
    
            Plugins are external binaries that Terraform uses to access and manipulate
            resources. The configuration provided requires plugins which can't be located,
            don't satisfy the version constraints, or are otherwise incompatible.
    
            Terraform automatically discovers provider requirements from your
            configuration, including providers used in child modules. To see the
            requirements and constraints from each module, run "terraform providers".
    
            - provider "awsalternate" is not available
    
            To match:
    
            .*peer_region cannot be set whilst auto_accept is true when creating a vpc peering connection.*
    
    --- FAIL: TestAccAWSVPCPeeringConnection_peerRegionAutoAccept (0.87s)
```

ewbankkit commented on aws/resource_aws_vpc_peering_connection_options_test.go:
```diff
@@ -213,28 +284,29 @@ func TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount(t *testing
-				ImportState:       true,
-				ImportStateVerify: true,
-			},
+			// Comment out until Plugin SDK correctly handles import testing with multiple providers.
```
Else get errors like

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount -timeout 120m
=== RUN   TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount
=== CONT  TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount
    testing.go:684: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
        
        (map[string]string) (len=4) {
         (string) (len=11) "requester.#": (string) (len=1) "1",
         (string) (len=44) "requester.0.allow_classic_link_to_remote_vpc": (string) (len=5) "false",
         (string) (len=43) "requester.0.allow_remote_vpc_dns_resolution": (string) (len=4) "true",
         (string) (len=44) "requester.0.allow_vpc_to_remote_classic_link": (string) (len=5) "false"
        }
        
        
        (map[string]string) (len=4) {
         (string) (len=10) "accepter.#": (string) (len=1) "1",
         (string) (len=43) "accepter.0.allow_classic_link_to_remote_vpc": (string) (len=5) "false",
         (string) (len=42) "accepter.0.allow_remote_vpc_dns_resolution": (string) (len=4) "true",
         (string) (len=43) "accepter.0.allow_vpc_to_remote_classic_link": (string) (len=5) "false"
        }
        
--- FAIL: TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount (33.37s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	33.453s
FAIL
GNUmakefile:26: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

I'll open a Plugin SDK issue & PR to fix this.

ewbankkit commented on aws/resource_aws_vpc_peering_connection_options_test.go:
```diff
@@ -213,28 +284,29 @@ func TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount(t *testing
 						"1",
 					),
-					tfawsresource.TestCheckTypeSetElemNestedAttrs(resourceNamePeer, "requester.*", map[string]string{
+					tfawsresource.TestCheckTypeSetElemNestedAttrs(resourceNamePeer, "accepter.*", map[string]string{
```
Fixes failing test case mentioned in #14383.

ewbankkit commented:
Rebased and simplified changes to reduce PR size.
Acceptance test output:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVPCPeeringConnection_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSVPCPeeringConnection_ -timeout 120m
=== RUN   TestAccAWSVPCPeeringConnection_basic
=== PAUSE TestAccAWSVPCPeeringConnection_basic
=== RUN   TestAccAWSVPCPeeringConnection_plan
=== PAUSE TestAccAWSVPCPeeringConnection_plan
=== RUN   TestAccAWSVPCPeeringConnection_tags
=== PAUSE TestAccAWSVPCPeeringConnection_tags
=== RUN   TestAccAWSVPCPeeringConnection_options
=== PAUSE TestAccAWSVPCPeeringConnection_options
=== RUN   TestAccAWSVPCPeeringConnection_failedState
=== PAUSE TestAccAWSVPCPeeringConnection_failedState
=== RUN   TestAccAWSVPCPeeringConnection_peerRegionAutoAccept
=== PAUSE TestAccAWSVPCPeeringConnection_peerRegionAutoAccept
=== RUN   TestAccAWSVPCPeeringConnection_region
=== PAUSE TestAccAWSVPCPeeringConnection_region
=== RUN   TestAccAWSVPCPeeringConnection_accept
=== PAUSE TestAccAWSVPCPeeringConnection_accept
=== RUN   TestAccAWSVPCPeeringConnection_optionsNoAutoAccept
=== PAUSE TestAccAWSVPCPeeringConnection_optionsNoAutoAccept
=== CONT  TestAccAWSVPCPeeringConnection_basic
=== CONT  TestAccAWSVPCPeeringConnection_peerRegionAutoAccept
--- PASS: TestAccAWSVPCPeeringConnection_peerRegionAutoAccept (14.09s)
=== CONT  TestAccAWSVPCPeeringConnection_optionsNoAutoAccept
--- PASS: TestAccAWSVPCPeeringConnection_basic (26.40s)
=== CONT  TestAccAWSVPCPeeringConnection_accept
--- PASS: TestAccAWSVPCPeeringConnection_optionsNoAutoAccept (16.97s)
=== CONT  TestAccAWSVPCPeeringConnection_region
--- PASS: TestAccAWSVPCPeeringConnection_region (25.91s)
=== CONT  TestAccAWSVPCPeeringConnection_options
--- PASS: TestAccAWSVPCPeeringConnection_accept (53.94s)
=== CONT  TestAccAWSVPCPeeringConnection_failedState
--- PASS: TestAccAWSVPCPeeringConnection_failedState (14.57s)
=== CONT  TestAccAWSVPCPeeringConnection_tags
--- PASS: TestAccAWSVPCPeeringConnection_options (41.97s)
=== CONT  TestAccAWSVPCPeeringConnection_plan
--- PASS: TestAccAWSVPCPeeringConnection_tags (25.91s)
--- PASS: TestAccAWSVPCPeeringConnection_plan (23.37s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	122.415s
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcPeeringConnectionOptions_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSVpcPeeringConnectionOptions_ -timeout 120m
=== RUN   TestAccAWSVpcPeeringConnectionOptions_basic
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_basic
=== RUN   TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccount
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccount
=== RUN   TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount
=== CONT  TestAccAWSVpcPeeringConnectionOptions_basic
=== CONT  TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount
--- PASS: TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount (32.41s)
=== CONT  TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccount
--- PASS: TestAccAWSVpcPeeringConnectionOptions_basic (44.41s)
--- PASS: TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccount (48.80s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	81.275s
```

DrFaust92 commented on aws/resource_aws_vpc_peering_connection.go:
```diff
@@ -254,10 +253,6 @@ func resourceAwsVPCPeeringUpdate(d *schema.ResourceData, meta interface{}) error
 		}
 	}
 
-	if err := vpcPeeringConnectionWaitUntilAvailable(conn, d.Id(), d.Timeout(schema.TimeoutUpdate)); err != nil {
```
no need to wait after modifying?

ewbankkit replied:
No, according to the [AWS documentation](https://docs.aws.amazon.com/vpc/latest/peering/vpc-peering-basics.html#vpc-peering-lifecycle), once `active` the state will only change to `deleting` or `deleted`.

DrFaust92 commented on aws/resource_aws_vpc_peering_connection_options.go:
```diff
@@ -78,14 +79,24 @@ func resourceAwsVpcPeeringConnectionOptionsUpdate(d *schema.ResourceData, meta i
+			req.RequesterPeeringConnectionOptions = expandVpcPeeringConnectionOptions(d.Get("requester").([]interface{}), crossRegionPeering)
+		}
+
+		log.Printf("[DEBUG] Modifying VPC Peering Connection options: %s", req)
```
same question regarding waiting here

ewbankkit replied:
See above.

ewbankkit replied:
In fact we do have to do some waiting after modification to deal with eventual consistency propagation of acceptance.

kenerwin88 commented:
I hope this gets merged in soon, need it badly :)

ewbankkit commented on aws/resource_aws_vpc_peering_connection_test.go:
```diff
@@ -432,7 +432,7 @@ func testAccCheckAWSVpcPeeringConnectionDestroy(s *terraform.State) error {
 
 		if pc.Status != nil {
-			if *pc.Status.Code == "deleted" || *pc.Status.Code == "rejected" {
+			if *pc.Status.Code == "deleted" || *pc.Status.Code == "rejected" || *pc.Status.Code == "failed" {
```
Else

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVPCPeeringConnection_failedState' ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSVPCPeeringConnection_failedState -timeout 120m
=== RUN   TestAccAWSVPCPeeringConnection_failedState
=== PAUSE TestAccAWSVPCPeeringConnection_failedState
=== CONT  TestAccAWSVPCPeeringConnection_failedState
    testing_new.go:62: Error running post-test destroy, there may be dangling resources: Found the VPC Peering Connection in an unexpected state: {
          AccepterVpcInfo: {
            OwnerId: "123456789012",
            Region: "us-west-2",
            VpcId: "vpc-05987a623d69770b3"
          },
          RequesterVpcInfo: {
            OwnerId: "123456789012",
            Region: "us-west-2",
            VpcId: "vpc-03951b70feaaed3bf"
          },
          Status: {
            Code: "failed",
            Message: "Failed due to incorrect VPC-ID, Account ID, or overlapping CIDR range"
          },
          Tags: [{
              Key: "Name",
              Value: "tf-testacc-pcx-39lbn0gn466xrouzb"
            }],
          VpcPeeringConnectionId: "pcx-0d8a9b6cada4e8510"
        }
--- FAIL: TestAccAWSVPCPeeringConnection_failedState (13.59s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	13.652s
FAIL
GNUmakefile:27: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

ewbankkit commented on aws/internal/service/ec2/waiter/status.go:
```diff
@@ -206,6 +207,43 @@ func SecurityGroupStatus(conn *ec2.EC2, id string) resource.StateRefreshFunc {
+)
+
+// VpcPeeringConnectionStatus fetches the VPC peering connection and its status
+func VpcPeeringConnectionStatus(conn *ec2.EC2, vpcPeeringConnectionID string) resource.StateRefreshFunc {
```
Can refactor existing VPC Peering Connection waiters to use this after merge.

ewbankkit replied:
Opened https://github.com/terraform-providers/terraform-provider-aws/issues/15908 to track this.

ewbankkit commented:
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVPCPeeringConnection_' ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSVPCPeeringConnection_ -timeout 120m
=== RUN   TestAccAWSVPCPeeringConnection_basic
=== PAUSE TestAccAWSVPCPeeringConnection_basic
=== RUN   TestAccAWSVPCPeeringConnection_plan
=== PAUSE TestAccAWSVPCPeeringConnection_plan
=== RUN   TestAccAWSVPCPeeringConnection_tags
=== PAUSE TestAccAWSVPCPeeringConnection_tags
=== RUN   TestAccAWSVPCPeeringConnection_options
=== PAUSE TestAccAWSVPCPeeringConnection_options
=== RUN   TestAccAWSVPCPeeringConnection_failedState
=== PAUSE TestAccAWSVPCPeeringConnection_failedState
=== RUN   TestAccAWSVPCPeeringConnection_peerRegionAutoAccept
=== PAUSE TestAccAWSVPCPeeringConnection_peerRegionAutoAccept
=== RUN   TestAccAWSVPCPeeringConnection_region
=== PAUSE TestAccAWSVPCPeeringConnection_region
=== RUN   TestAccAWSVPCPeeringConnection_accept
=== PAUSE TestAccAWSVPCPeeringConnection_accept
=== RUN   TestAccAWSVPCPeeringConnection_optionsNoAutoAccept
=== PAUSE TestAccAWSVPCPeeringConnection_optionsNoAutoAccept
=== CONT  TestAccAWSVPCPeeringConnection_basic
=== CONT  TestAccAWSVPCPeeringConnection_peerRegionAutoAccept
--- PASS: TestAccAWSVPCPeeringConnection_peerRegionAutoAccept (14.54s)
=== CONT  TestAccAWSVPCPeeringConnection_failedState
--- PASS: TestAccAWSVPCPeeringConnection_failedState (13.27s)
=== CONT  TestAccAWSVPCPeeringConnection_options
--- PASS: TestAccAWSVPCPeeringConnection_basic (35.23s)
=== CONT  TestAccAWSVPCPeeringConnection_tags
--- PASS: TestAccAWSVPCPeeringConnection_options (62.47s)
=== CONT  TestAccAWSVPCPeeringConnection_plan
--- PASS: TestAccAWSVPCPeeringConnection_plan (24.99s)
=== CONT  TestAccAWSVPCPeeringConnection_optionsNoAutoAccept
--- PASS: TestAccAWSVPCPeeringConnection_tags (86.90s)
=== CONT  TestAccAWSVPCPeeringConnection_accept
=== CONT  TestAccAWSVPCPeeringConnection_region
--- PASS: TestAccAWSVPCPeeringConnection_optionsNoAutoAccept (14.51s)
--- PASS: TestAccAWSVPCPeeringConnection_region (36.92s)
--- PASS: TestAccAWSVPCPeeringConnection_accept (90.20s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	212.871s
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcPeeringConnectionOptions_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSVpcPeeringConnectionOptions_ -timeout 120m
=== RUN   TestAccAWSVpcPeeringConnectionOptions_basic
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_basic
=== RUN   TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccount
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccount
=== RUN   TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount
=== PAUSE TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount
=== CONT  TestAccAWSVpcPeeringConnectionOptions_basic
=== CONT  TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount
=== CONT  TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccount
--- PASS: TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount (46.81s)
--- PASS: TestAccAWSVpcPeeringConnectionOptions_basic (61.15s)
--- PASS: TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccount (73.24s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	73.598s
```

bflad reviewed (approved):
Looks good, thank you for rebasing and updating 🚀 

Output from acceptance testing:

```
--- PASS: TestAccAWSVPCPeeringConnection_accept (91.89s)
--- PASS: TestAccAWSVPCPeeringConnection_basic (48.18s)
--- PASS: TestAccAWSVPCPeeringConnection_failedState (23.66s)
--- PASS: TestAccAWSVPCPeeringConnection_options (72.09s)
--- PASS: TestAccAWSVPCPeeringConnection_optionsNoAutoAccept (29.30s)
--- PASS: TestAccAWSVPCPeeringConnection_peerRegionAutoAccept (28.12s)
--- PASS: TestAccAWSVPCPeeringConnection_plan (43.81s)
--- PASS: TestAccAWSVPCPeeringConnection_region (62.99s)
--- PASS: TestAccAWSVPCPeeringConnection_tags (89.98s)

--- PASS: TestAccAWSVpcPeeringConnectionOptions_basic (48.74s)
--- PASS: TestAccAWSVpcPeeringConnectionOptions_differentRegionSameAccount (55.67s)
--- PASS: TestAccAWSVpcPeeringConnectionOptions_sameRegionDifferentAccount (34.38s)
```
